### PR TITLE
Add struct data type front end

### PR DESCRIPTION
**Pull request by acquamarin**
1. Implement front end for `STRUCT` data type. Grammar for `STRUCT` type is: STRUCT(prop1 prop1Type, prop2 prop2Type). 
2. Renames `propertyIdx` -> `propertyID`.
3. Refactors how dataType stores extraTypeInfo.

### CODE REVIEW

File: src/storage/copier/rel_copier.cpp [hunk 1]
```diff
@@ -392,70 +392,70 @@ void RelCopier::putPropsOfLineIntoColumns(RelCopier* copier,
...
         case INT16: {
             auto val = TypeUtils::convertStringToNumber<int16_t>(data);
-            putValueIntoColumns(propertyIdx, directionTablePropertyColumns, nodeIDs,
+            putValueIntoColumns(propertyID, directionTablePropertyColumns, nodeIDs,
                 reinterpret_cast<uint8_t*>(&val));
         } break;
         case DOUBLE: {
             auto val = TypeUtils::convertStringToNumber<double_t>(data);
-            putValueIntoColumns(propertyIdx, directionTablePropertyColumns, nodeIDs,
+            putValueIntoColumns(propertyID, directionTablePropertyColumns, nodeIDs,
```
Comment (andyfengHKU) [inline src/storage/copier/rel_copier.cpp:413]: hmmm should we open an issue about revisit test coverage at some point? Seems we don't have test for DOUBLE rel property and BOOL rel property

File: test/binder/binder_error_test.cpp [hunk 2]
```diff
@@ -489,3 +489,16 @@ TEST_F(BinderErrorTest, InvalidFixedListSize) {
     auto input = "create node table test1(ID INT64, marks INT64[512], PRIMARY KEY(ID))";
     ASSERT_STREQ(expectedException.c_str(), getBindingError(input).c_str());
 }
+
+TEST_F(BinderErrorTest, MissingStructFieldType) {
```
Comment (andyfengHKU) [inline test/binder/binder_error_test.cpp:493]: If you are changing grammar, don't forget to remove these 2 tests in your next PR

**Review by andyfengHKU — APPROVED** [review]

File: test/copy/copy_npy_test.cpp [hunk 3]
```diff
@@ -48,33 +48,33 @@ TEST_F(CopyOneDimensionalNpyTest, CopyOneDimensionalNpyTest) {
     auto storageManager = getStorageManager(*database);
     auto catalog = getCatalog(*database);
     auto tableID = catalog->getReadOnlyVersion()->getTableID("npytable");
-    auto propertyIdx = catalog->getReadOnlyVersion()->getNodeProperty(tableID, "i64");
+    auto propertyID = catalog->getReadOnlyVersion()->getNodeProperty(tableID, "i64");
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply test/copy/copy_npy_test.cpp:51]: Renamed to property

File: src/common/types/types.cpp [hunk 4]
```diff
@@ -1,63 +1,182 @@
...
+
+template<>
+uint64_t SerDeser::deserializeValue(DataType& value, FileInfo* fileInfo, uint64_t offset) {
+    offset = SerDeser::deserializeValue(value.typeID, fileInfo, offset);
+    switch (value.getTypeID()) {
+    case VAR_LIST: {
+        value.extraTypeInfo = std::make_unique<VarListTypeInfo>();
+        offset = deserializeValue(
+            *reinterpret_cast<VarListTypeInfo*>(value.getExtraTypeInfo()), fileInfo, offset);
+        break;
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/common/types/types.cpp:87]: Changed code style

File: src/storage/copier/rel_copier.cpp (hunk 1 above)
Comment (acquamarin) [reply src/storage/copier/rel_copier.cpp:413]: i think it is ok not to test double/bool rel property. If we test each datatype on node column, rel column, rel lists, our testing code will explode.

File: src/common/types/types.cpp [hunk 5]
```diff
@@ -1,63 +1,182 @@
...
+    return *childType == *other.childType && fixedNumElementsInList == other.fixedNumElementsInList;
+}
 
-DataType::DataType(DataTypeID typeID) : typeID{typeID}, childType{nullptr} {
-    assert(typeID != VAR_LIST && typeID != FIXED_LIST);
+std::unique_ptr<ExtraTypeInfo> FixedListTypeInfo::copy() const {
+    return std::make_unique<FixedListTypeInfo>(childType->copy(), fixedNumElementsInList);
 }
 
+bool StructTypeInfo::operator==(const kuzu::common::StructTypeInfo& other) const {
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/common/types/types.cpp:124]: replaced with assert false.

**Pull request merged**